src/object/cli_obj.c

```diff
@@ -2645,3 +2662,540 @@ dc_obj_query_key(tse_task_t *api_task)
+				       "%d (l) does not exists, but shard %d "
+				       "does exists.\n", DP_OID(oid), reps,
+				       leader + start, i + start);
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+
+			if (arg_s->lost_shard) {
+				/* Non-leader does not exist, but leader seems
+				 * there. It may because someone destroyed the
+				 * object after we list from leader. Let's try
+				 * to re-list from leader.
+				 */
+				if (!arg_l->has_retried &&
+				    dc_obj_need_recheck(oeda, reps, start,
+							leader, i, flags,
+							"re-list_2"))
+					goto list;
+
+				D_INFO(DF_OID" (reps %d, inconsistent) shard "
+				       "%d (l) does exists, but shard %d does "
+				       "not exists.\n", DP_OID(oid), reps,
+				       leader + start, i + start);
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+
+			if (arg_l->size != arg_s->size ||
+			    memcmp(arg_l->buf, arg_s->buf, arg_l->size) != 0) {
+				/* Non-leader has different keys. It may because
+				 * someone changed the object after we list from
+				 * leader. Let's re-list from leader.
+				 */
+				if (!arg_l->has_retried &&
+				    dc_obj_need_recheck(oeda, reps, start,
+							leader, i, flags,
+							"re-list_3"))
+					goto list;
+
+				D_INFO(DF_OID" (reps %d, inconsistent) shard "
+				       "%d (l) and shard %d have inconsistent "
+				       "keys/recs list\n", DP_OID(oid), reps,
+				       leader + start, i + start);
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+		}
+
+		if (arg_l->lost_shard || arg_l->num == 0)
+			break;
+
+		daos_enum_dkeys_prep_unpack(arg_l);
+		rc = obj_enum_unpack(&arg_l->enum_arg,
+				     dc_obj_verify_one_rdg_unpack_cb,
+				     &unpack_arg);
+		if (rc != 0)
+			break;
+
+		if (daos_anchor_is_eof(&arg_l->dkey_anchor))
+			break;
+	}
+
+out:
+	if (oeda != NULL) {
+		for (i = 0; i < reps; i++)
+			daos_enum_dkeys_fini_arg(&oeda[i]);
+		D_FREE(oeda);
+	}
+
+	return rc;
+}
+
+static int
+dc_obj_verify_list_wrap(daos_handle_t oh, daos_epoch_t *epoch, daos_key_t *dkey,
+			daos_key_t *akey, daos_size_t *size, uint32_t *nr,
+			daos_key_desc_t *kds, daos_epoch_range_t *eprs,
+			d_sg_list_t *sgl, daos_anchor_t *anchor,
+			daos_anchor_t *dkey_anchor, daos_anchor_t *akey_anchor)
+{
+	tse_task_t	*task;
+	int		 rc;
+
+	rc = dc_obj_list_obj_task_create(oh, DAOS_TX_NONE, epoch, dkey, akey,
+					 size, nr, kds, eprs, sgl, anchor,
+					 dkey_anchor, akey_anchor, true,
+					 NULL, NULL, &task);
+	if (rc == 0)
+		rc = dc_task_schedule(task, true);
+
+	return rc;
+}
+
+int
+dc_obj_verify(daos_handle_t oh, uint32_t flags)
+{
+	struct dc_object	*obj;
+	unsigned int		 replicas;
+	int			 rc = 0;
+	int			 i;
+
+	/* XXX: Currently, we always assume that there is no concurrent
+	 *	modification during the verification. So if we got some
+	 *	different things from different replicas, then it should
+	 *	be caused by the real inconsistency. It is the caller's
+	 *	duty to guarantee the correctness of above assumption.
+	 */
+	D_ASSERT(flags & DOVF_IMMUTABLE);
```

Comment:
So looks a "VERIFY_EXCLUSIVE" matches the comment better.

Reply:
Hm, EXCLUSIVE means there is no other users to access the target, but the current user can modify the target by itself. IMMUTABLE means that nobody will modify the target in spite of how many users. The latter one is what we need.